halo-dev/halo
# Implement simple garbage collector to collect deletable extensions

#### What type of PR is this?

/kind feature
/area core
/milestone 2.0

#### What this PR does / why we need it:

- Add finalizers field into metadata to let controller do some works before deleting
- Implement a simple garbage collector to collect deletable extensions

Garbage collector controller is a special controller, which will watch extensions with any GVK instead of specific type. So we have to customize the controller parameters entirely.

#### Which issue(s) this PR fixes:

Fixes #2307

#### Special notes for your reviewer:

How to test?

- Delete without finalizers

    1. Create an extension and check it
    2. Delete it and check it again

- Delete with finalizers
    
    1. Create an extension and update it with finalizers
    2. Delete it and checkout it
    4. You will see the extension with finalizers not deleted
    5. Update it without finalizers and check it again

#### Does this PR introduce a user-facing change?

```release-note
None
```
